qiskit/circuit/library/arithmetic/functional_pauli_rotations.py

```diff
@@ -0,0 +1,137 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Base class for functional Pauli rotations."""
+
+from typing import Optional, List, Tuple
+
+from abc import ABC, abstractmethod
+from qiskit.circuit import QuantumCircuit, Qubit, Clbit, Instruction
+
+
+class FunctionalPauliRotations(QuantumCircuit, ABC):
+    """Base class for functional Pauli rotations."""
+
+    def __init__(self,
+                 num_state_qubits: Optional[int] = None,
+                 basis: str = 'Y',
+                 name: str = 'F') -> None:
+        r"""Create a new functional Pauli rotation circuit.
+
+        Args:
+            num_state_qubits: The number of qubits representing the state :math:`|x\rangle`.
+            basis: The kind of Pauli rotation to use. Must be 'X', 'Y' or 'Z'.
+            name: The name of the circuit object.
+        """
+        super().__init__(name=name)
+
+        # define internal parameters
+        self._num_state_qubits = None
+        self._basis = None
+
+        # store parameters
+        self.num_state_qubits = num_state_qubits
+        self.basis = basis
+
+    @property
+    def basis(self) -> str:
+        """The kind of Pauli rotation to be used.
+
+        Set the basis to 'X', 'Y' or 'Z' for controlled-X, -Y, or -Z rotations respectively.
+
+        Returns:
+            The kind of Pauli rotation used in controlled rotation.
+        """
+        return self._basis
+
+    @basis.setter
+    def basis(self, basis: str) -> None:
+        """Set the kind of Pauli rotation to be used.
+
+        Args:
+            basis: The Pauli rotation to be used.
+
+        Raises:
+            ValueError: The provided basis in not X, Y or Z.
+        """
+        basis = basis.lower()
+        if self._basis is None or basis != self._basis:
+            if basis not in ['x', 'y', 'z']:
+                raise ValueError('The provided basis must be X, Y or Z, not {}'.format(basis))
+            self._basis = basis
+            self._data = None
+
+    @property
+    def num_state_qubits(self) -> int:
+        r"""The number of state qubits representing the state :math:`|x\rangle`.
+
+        Returns:
+            The number of state qubits.
+        """
+        return self._num_state_qubits
+
+    @num_state_qubits.setter
+    def num_state_qubits(self, num_state_qubits: Optional[int]) -> None:
```

Comment:
So that the circuit can be "reset" by 
```
circuit.num_state_qubits = None
```
Maybe an application checks of the number of qubits has been set or not, though I don't have a concrete example where this would be needed. Should I set it to not support `None`?

Reply:
that's fine. i just see that in this PR some setters accept `None` and some don't, is that intentional?